File: frontend/lib/models/notification.dart
import 'package:flutter/material.dart';
import 'package:json_annotation/json_annotation.dart';

part 'notification.g.dart';

@JsonSerializable()
class Notification {
  final String id;
  final String type;
  final String message;
  final String recipient;
  final bool read;
  final String? ticketId;
  final DateTime createdAt;
  final Map<String, dynamic>? metadata;

  Notification({
    required this.id,
    required this.type,
    required this.message,
    required this.recipient,
    required this.read,
    this.ticketId,
    required this.createdAt,
    this.metadata,
  });

  factory Notification.fromJson(Map<String, dynamic> json) =>
      _$NotificationFromJson(json);

  Map<String, dynamic> toJson() => _$NotificationToJson(this);

  NotificationType get notificationType =>
      NotificationTypeExtension.fromString(type);

  String get typeIcon {
    switch (notificationType) {
      case NotificationType.ticketAssigned:
        return 'assignment';
      case NotificationType.slaBreached:
        return 'warning';
      case NotificationType.escalation:
        return 'arrow_upward';
      case NotificationType.shiftEnding:
        return 'access_time';
      case NotificationType.handover:
        return 'swap_horiz';
      case NotificationType.breakReminder:
        return 'coffee';
      default:
        return 'notifications';
    }
  }

  Color get typeColor {
    switch (notificationType) {
      case NotificationType.ticketAssigned:
        return Colors.blue;
      case NotificationType.slaBreached:
        return Colors.red;
      case NotificationType.escalation:
        return Colors.orange;
      case NotificationType.shiftEnding:
        return Colors.purple;
      case NotificationType.handover:
        return Colors.green;
      case NotificationType.breakReminder:
        return Colors.teal;
      default:
        return Colors.grey;
    }
  }
}

enum NotificationType {
  ticketAssigned,
  slaBreached,
  escalation,
  shiftEnding,
  handover,
  breakReminder,
  other
}

extension NotificationTypeExtension on NotificationType {
  static NotificationType fromString(String type) {
    switch (type) {
      case 'ticket_assigned':
        return NotificationType.ticketAssigned;
      case 'sla_breach':
        return NotificationType.slaBreached;
      case 'escalation':
        return NotificationType.escalation;
      case 'shift_ending':
        return NotificationType.shiftEnding;
      case 'handover':
        return NotificationType.handover;
      case 'break_reminder':
        return NotificationType.breakReminder;
      default:
        return NotificationType.other;
    }
  }
}

File: backend/ticket-system-backend/src/models/Notification.js
const mongoose = require("mongoose");

const notificationSchema = new mongoose.Schema(
	{
		type: {
			type: String,
			required: true,
			enum: [
				"ticket_assigned",
				"sla_breach",
				"escalation",
				"shift_ending",
				"handover",
				"break_reminder",
			],
		},
		recipient: {
			type: mongoose.Schema.Types.ObjectId,
			ref: "User",
			required: true,
		},
		message: {
			type: String,
			required: true,
		},
		read: {
			type: Boolean,
			default: false,
		},
		ticket: {
			type: mongoose.Schema.Types.ObjectId,
			ref: "Ticket",
		},
		metadata: {
			type: mongoose.Schema.Types.Mixed,
		},
	},
	{
		timestamps: true,
	}
);

notificationSchema.index({ recipient: 1, read: 1, createdAt: -1 });
notificationSchema.index({ type: 1, createdAt: -1 });

module.exports = mongoose.model("Notification", notificationSchema);
